pkg/generator/cloudsmith/cloudsmith.go

```diff
@@ -0,0 +1,203 @@
+)
+
+type Generator struct {
+	httpClient *http.Client
+}
+
+type CloudsmithOIDCRequest struct {
+	OIDCToken   string `json:"oidc_token"`
+	ServiceSlug string `json:"service_slug"`
+}
+
+type CloudsmithOIDCResponse struct {
+	Token string `json:"token"`
+}
+
+const (
+	defaultCloudsmithAPIHost = "api.cloudsmith.io"
+
+	errNoSpec            = "no config spec provided"
+	errParseSpec         = "unable to parse spec: %w"
+	errGetToken          = "unable to get authorization token: %w"
+	errExchangeToken     = "unable to exchange OIDC token: %w"
+	errMarshalRequest    = "failed to marshal request payload: %w"
+	errCreateRequest     = "failed to create HTTP request: %w"
+	errUnexpectedStatus  = "request failed due to unexpected status: %s"
+	errReadResponse      = "failed to read response body: %w"
+	errUnmarshalResponse = "failed to unmarshal response: %w"
+	errTokenNotFound     = "token not found in response"
+	errTokenTypeCast     = "error when typecasting token to string"
+
+	httpClientTimeout = 30 * time.Second
+)
+
+func (g *Generator) Generate(ctx context.Context, cloudsmithSpec *apiextensions.JSON, kubeClient client.Client, targetNamespace string) (map[string][]byte, genv1alpha1.GeneratorProviderState, error) {
+	return g.generate(
+		ctx,
+		cloudsmithSpec,
+		kubeClient,
+		targetNamespace,
+	)
+}
+
+func (g *Generator) Cleanup(_ context.Context, cloudsmithSpec *apiextensions.JSON, providerState genv1alpha1.GeneratorProviderState, _ client.Client, _ string) error {
+	return nil
+}
+
+func (g *Generator) generate(
+	ctx context.Context,
+	cloudsmithSpec *apiextensions.JSON,
+	_ client.Client,
+	targetNamespace string) (map[string][]byte, genv1alpha1.GeneratorProviderState, error) {
+	if cloudsmithSpec == nil {
+		return nil, nil, errors.New(errNoSpec)
+	}
+	res, err := parseSpec(cloudsmithSpec.Raw)
+	if err != nil {
+		return nil, nil, fmt.Errorf(errParseSpec, err)
+	}
+
+	// Fetch the service account token
+	oidcToken, err := utils.FetchServiceAccountToken(ctx, res.Spec.ServiceAccountRef, targetNamespace)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch service account token: %w", err)
+	}
+
+	apiHost := res.Spec.APIHost
+	if apiHost == "" {
+		apiHost = defaultCloudsmithAPIHost
+	}
+	apiHost = strings.TrimPrefix(apiHost, "https://")
+
+	accessToken, err := g.exchangeTokenWithCloudsmith(ctx, oidcToken, res.Spec.OrgSlug, res.Spec.ServiceSlug, apiHost)
+	if err != nil {
+		return nil, nil, fmt.Errorf(errExchangeToken, err)
+	}
+
+	exp, err := utils.ExtractJWTExpiration(accessToken)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return map[string][]byte{
+		"auth": []byte(b64.StdEncoding.EncodeToString([]byte("token:" + accessToken))),
+		"expiry": []byte(exp),
+	}, nil, nil
+}
+
+func (g *Generator) exchangeTokenWithCloudsmith(ctx context.Context, oidcToken, orgSlug, serviceSlug, apiHost string) (string, error) {
+	klog.V(4).InfoS("Starting OIDC token exchange with Cloudsmith")
```

Comment:
Please don't use klog. Try fetching the logger from context ( it should be available since it's passed through the manager ). And if it doesn't exist use the control-runtime logger. There are examples of this here and there.